src/platform/posix/tcp_ip_channel.c

```diff
@@ -411,41 +457,56 @@ static void *_TcpIpChannel_worker_thread(void *untyped_self) {
-      } else {
-        _TcpIpChannel_try_connect_client(untyped_self);
-      }
-    } break;
-
-    case NETWORK_CHANNEL_STATE_CONNECTION_IN_PROGRESS: {
-      _env->platform->wait_for(_env->platform, self->super.expected_try_connect_duration);
-    } break;
-
-    case NETWORK_CHANNEL_STATE_LOST_CONNECTION:
-    case NETWORK_CHANNEL_STATE_CONNECTION_FAILED: {
-      _env->platform->wait_for(_env->platform, self->super.expected_try_connect_duration);
-      _TcpIpChannel_reset_socket(self);
-      _TcpIpChannel_update_state(self, NETWORK_CHANNEL_STATE_OPEN);
-    } break;
-
-    case NETWORK_CHANNEL_STATE_CONNECTED: {
-      ret = _TcpIpChannel_receive(untyped_self, &self->output);
-      if (ret == LF_OK) {
-        validate(self->receive_callback);
-        self->receive_callback(self->federated_connection, &self->output);
-      } else {
-        LF_ERR(NET, "TcpIpChannel: Error receiving message %d", ret);
+    pthread_mutex_lock(&self->state_mutex);
+    {
+      switch (_TcpIpChannel_get_state(self)) {
+      case NETWORK_CHANNEL_STATE_OPEN: {
+        /* try to connect */
+        if (self->server) {
+          _TcpIpChannel_try_connect_server(untyped_self);
+        } else {
+          _TcpIpChannel_try_connect_client(untyped_self);
+        }
+      } break;
+
+      case NETWORK_CHANNEL_STATE_CONNECTION_IN_PROGRESS: {
+        _env->platform->wait_for(_env->platform, self->super.expected_connect_duration);
+      } break;
+
+      case NETWORK_CHANNEL_STATE_LOST_CONNECTION:
+      case NETWORK_CHANNEL_STATE_CONNECTION_FAILED: {
+        _env->platform->wait_for(_env->platform, self->super.expected_connect_duration);
+        _TcpIpChannel_reset_socket(self);
+        _TcpIpChannel_update_state(self, NETWORK_CHANNEL_STATE_OPEN);
+      } break;
+
+      case NETWORK_CHANNEL_STATE_CONNECTED: {
+        int socket = 0;
+        if (self->server) {
+          socket = self->client;
+        } else {
+          socket = self->fd;
+        }
+        _TcpIpChannel_socket_configure_blocking(socket, false);
+        ret = _TcpIpChannel_receive(untyped_self, &self->output);
+        _TcpIpChannel_socket_configure_blocking(socket, true);
+        if (ret == LF_EMPTY) {
+          /* The non-blocking recv has no new data yet */
+        } else if (ret == LF_OK) {
+          validate(self->receive_callback);
+          self->receive_callback(self->federated_connection, &self->output);
+        } else if (ret == LF_ERR) {
+          LF_ERR(NET, "TcpIpChannel: Error receiving message %d", ret);
+        }
+      } break;
+
+      case NETWORK_CHANNEL_STATE_UNINITIALIZED:
+      case NETWORK_CHANNEL_STATE_CLOSED:
+        break;
       }
-    } break;
-
-    case NETWORK_CHANNEL_STATE_UNINITIALIZED:
-    case NETWORK_CHANNEL_STATE_CLOSED:
-      break;
     }
+    pthread_mutex_unlock(&self->state_mutex);
+    _env->platform->wait_for(_env->platform, TCP_IP_CHANNEL_WORKER_THREAD_MAIN_LOOP_SLEEP);
```

Comment:
It is a little unfortunate that we have to set a sleep here. The duration of the sleep adds to the worst case latency between two federates. Also it means that we are always polling the the receive socket and that we never really can go to low-power sleep